### ISSUE TITLE
`reportMissingSource` is inconsistent under stub/implementation directory mismatches

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/pyright#6097]
**Describe the bug**

When using `reportMissingSource`, reporting an error seems to depend on where the stub and implementation live relative to each other.

**Code or Screenshots**

```python
# subdir/foo.py

import blah

blah.yadda()

# subdir/bar.py

def yadda() -> None:
    raise NotImplementedError()

# bar.pyi

def yadda() -> None:
    ...
```

When using `reportMissingSource`, Pyright reports

```
Import "blah" could not be resolved from source
```

However, if you swap the parent directory of the `bar.py` and `bar.pyi`:

* `subdir/bar.py` ➡️ `bar.py`
* `bar.pyi` ➡️ `subdir/bar.pyi`

Then this error goes away.

**VS Code extension or command-line**

Pylance 2023.10.10

**Comment by erictraut** [comment]
I presume that the `subdir/foo.py` code sample is supposed to say `'import bar` and `bar.yadda()`. If I make that change, pyright reports a `reportMissingSource` diagnostic as expected. That's because the import is resolved using the `bar.pyi` source file, which is at the root of the project. The `bar.py` source isn't available to resolve the import because it's in a subdirectory, and you haven't configured pyright to use an "extraPaths" that includes "subdir".

If I swap the parent directory of `bar.py` and `bar.pyi`, then the `import bar` statement is resolved via the `bar.py` source file.

So this is all working as intended.
